xdsl/dialects/scf.py

```diff
@@ -559,79 +555,78 @@ def verify_(self) -> None:
         # scf.parallel result type (there is no result type on scf.reduce, hence we check the
         # operand type)
         for reduction in range(num_reductions):
-            arg_type = self.get_arg_type_of_nth_reduction_op(reduction)
+            arg_type = reduce_op.args[reduction].type
             res_type = self.res[reduction].type
             if res_type != arg_type:
                 raise VerifyException(
                     f"Miss match on scf.parallel result type and reduction op type number {reduction} "
                     f", parallel argment is of type {res_type} whereas reduction operation is of type {arg_type}"
                 )
 
-    def count_number_reduction_ops(self) -> int:
-        num_reduce = 0
-        for op in self.body.block.ops:
-            if isinstance(op, ReduceOp):
-                num_reduce += 1
-        return num_reduce
-
-    def get_arg_type_of_nth_reduction_op(self, index: int):
-        found = 0
-        for op in self.body.block.ops:
-            if isinstance(op, ReduceOp):
-                if found == index:
-                    return op.argument.type
-                found += 1
-        return None
-
 
 @irdl_op_definition
 class ReduceOp(IRDLOperation):
     name = "scf.reduce"
-    argument = operand_def()
+    args = var_operand_def()
 
-    body = region_def("single_block")
+    reductions = var_region_def("single_block")
 
-    traits = frozenset([RecursiveMemoryEffect()])
+    traits = traits_def(
+        lambda: frozenset(
+            [
+                RecursiveMemoryEffect(),
+                HasParent(ParallelOp),
+                IsTerminator(),
+                SingleBlockImplicitTerminator(ReduceReturnOp),
+            ]
+        )
+    )
 
     def __init__(
         self,
-        argument: SSAValue | Operation,
-        block: Block,
+        args: Sequence[SSAValue | Operation] = (),
+        blocks: Sequence[Block] = (),
```

Comment:
it feels to me like this should take a sequence of regions, to a void a situation where you have a list of regions, and need to create a tuple of first blocks, only to then create new regions with those blocks.